pkg/cloudprovider/aws/capacity.go

```diff
@@ -35,34 +49,57 @@ type Capacity struct {
 	instanceTypeProvider   *InstanceTypeProvider
 }
 
+// AWSConstraints are AWS specific constraints
+type AWSConstraints struct {
+	cloudprovider.Constraints
+	CapacityType string
+}
+
+func NewAWSConstraints(constraints *cloudprovider.Constraints) *AWSConstraints {
+	capacityType, ok := constraints.Labels[capacityTypeLabel]
+	if !ok {
+		capacityType = capacityTypeOnDemand
+	}
+	return &AWSConstraints{
+		Constraints:  *constraints,
+		CapacityType: capacityType,
+	}
+}
+
 // Create a set of nodes given the constraints.
 func (c *Capacity) Create(ctx context.Context, constraints *cloudprovider.Constraints) ([]cloudprovider.Packing, error) {
-	// 1. Retrieve normalized zones from constraints or all zones that the cluster spans if no zonal constraints are specified
-	zonalSubnetOptions, err := c.vpcProvider.GetZonalSubnets(ctx, constraints, c.spec.Cluster.Name)
+	// 1. Create AWS Cloud Provider constraints and apply defaults
+	awsConstraints := NewAWSConstraints(constraints)
+
+	// 2. Retrieve normalized zones from constraints or all zones that the cluster spans if no zonal constraints are specified
+	zonalSubnetOptions, err := c.vpcProvider.GetZonalSubnets(ctx, awsConstraints, c.spec.Cluster.Name)
 	if err != nil {
 		return nil, fmt.Errorf("getting zonal subnets, %w", err)
 	}
 
-	// 2. Filter for instance types that fit constraints
-	zonalInstanceTypes, err := c.instanceTypeProvider.Get(ctx, zonalSubnetOptions, constraints)
+	// 3. Filter for instance types that fit constraints
+	zonalInstanceTypes, err := c.instanceTypeProvider.Get(ctx, zonalSubnetOptions, awsConstraints)
 	if err != nil {
 		return nil, fmt.Errorf("filtering instance types by constraints, %w", err)
 	}
 
-	// 3. Compute Packing given the pods and instance types
-	instancePackings := c.packer.Pack(ctx, constraints.Pods, zonalInstanceTypes, constraints)
+	// 4. Compute Packing given the pods and instance types
+	instancePackings := c.packer.Pack(ctx, awsConstraints.Pods, zonalInstanceTypes, constraints)
 
-	zap.S().Debugf("Computed packings for %d pod(s) onto %d node(s)", len(constraints.Pods), len(instancePackings))
-	launchTemplate, err := c.launchTemplateProvider.Get(ctx, c.spec.Cluster, constraints)
+	zap.S().Debugf("Computed packings for %d pod(s) onto %d node(s)", len(awsConstraints.Pods), len(instancePackings))
+	launchTemplate, err := c.launchTemplateProvider.Get(ctx, c.spec.Cluster, awsConstraints)
 	if err != nil {
 		return nil, fmt.Errorf("getting launch template, %w", err)
 	}
 
-	// 4. Create Instances
+	// 5. Create Instances
 	var instanceIDs []*string
 	podsForInstance := make(map[string][]*v1.Pod)
 	for _, packing := range instancePackings {
-		instanceID, err := c.instanceProvider.Create(ctx, launchTemplate, packing.InstanceTypes, zonalSubnetOptions)
+		if len(packing.InstanceTypes) > maxInstanceTypes {
```

Comment:
imo this limiting detail should be encapsulated by the instanceProvider class

Reply:
I had originally thought that as well, but I'm not sure it makes sense there since that would require us to trim the list before pods are packed which means the packer has less options to choose how to pack pods.  

I also considered the packer to do the limiting, but didn't think it made sense there either since it might be cloud provider specific if you even want to limit packing options.

Reply:
>but I'm not sure it makes sense there since that would require us to trim the list before pods are packed which means the packer has less options to choose how to pack pods.
I'm not sure I agree. The next line is c.instanceProvider.Create(..., instancetype, ...), surely it could happen inside there without impacting the flow?

Reply:
oh sorry, I read that wrong, thought you were suggesting the instanceTypeProvider... 

I guess I'm neutral on the instanceProvider doing the trimming. I thought it might be surprising to have instanceTypes truncated when passed a list of options. But I suppose it only has one use-case and we don't want it to use all of them so... I guess it doesn't really matter.  As I'm typing I'm coming around to the idea... :)